File: backend/app/routers/brain.py
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.db_models import User
from app.deps import get_current_user
from app.services.brain_service import (
    get_brain_feed,
    get_brain_status,
    get_daily_brief,
    get_weekly_report,
    ingest_url,
    learn_today,
    list_agents,
    morning_cycle,
    run_agent,
    run_all_active_agents,
)
from app.services.brain_config import config_to_dict, get_brain_config, update_brain_config
from app.services.brain_scheduler import run_cron_for_all_users
from app.services.business_context import build_business_context

router = APIRouter(prefix="/api/v1/brain", tags=["brain"])


class IngestUrlRequest(BaseModel):
    url: str


class BrainConfigPatch(BaseModel):
    competitors: list[str] | None = None
    brand_keywords: list[str] | None = None
    auto_run_daily: bool | None = None
    enabled_agents: list[str] | None = None


@router.get("/status")
async def brain_status(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await get_brain_status(db, user)


@router.get("/agents")
async def brain_agents(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    context = await build_business_context(user, db)
    return {"agents": list_agents(context)}


@router.get("/daily-brief")
async def daily_brief(
    refresh: bool = Query(False),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await get_daily_brief(db, user, refresh=refresh)


@router.get("/weekly-report")
async def weekly_report(
    refresh: bool = Query(False),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await get_weekly_report(db, user, refresh=refresh)


@router.get("/feed")
async def brain_feed(
    limit: int = Query(30, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return {"items": await get_brain_feed(db, user, limit=limit)}


@router.post("/learn")
async def brain_learn(
    force: bool = Query(False),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    learned = await learn_today(db, user, force=force)
    return {"ok": True, "learned": learned}


@router.post("/ingest-url")
async def brain_ingest_url(
    body: IngestUrlRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await ingest_url(db, user, body.url)


@router.post("/agents/{agent_id}/run")
async def brain_run_agent(
    agent_id: str,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await run_agent(db, user, agent_id)


@router.post("/run-all")
async def brain_run_all(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await run_all_active_agents(db, user)


@router.post("/morning-cycle")
async def brain_morning_cycle(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await morning_cycle(db, user)


@router.get("/config")
async def brain_get_config(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    row = await get_brain_config(db, user.id)
    return config_to_dict(row)


@router.patch("/config")
async def brain_patch_config(
    body: BrainConfigPatch,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    patch = body.model_dump(exclude_none=True)
    return await update_brain_config(db, user.id, patch)


@router.post("/cron")
async def brain_cron(secret: str = Query("")):
    return await run_cron_for_all_users(secret)

File: frontend/src/lib/brain.ts
import { apiFetch } from "./api";

export type BrainSource = {
  id: string;
  label: string;
  type: string;
};

export type BrainAgent = {
  id: string;
  tag: string;
  name: string;
  description: string;
  outcome: string;
  integrations: string[];
  status: "active" | "needs_setup";
  schedule?: string;
};

export type BrainStatus = {
  company: string;
  goal: string;
  tagline: string;
  days_learned: number;
  memory_entries: number;
  learned_today: boolean;
  agents_active: number;
  agents_total: number;
  agents_run_today: number;
  sources: BrainSource[];
  connected_integrations: string[];
  narrative: string;
  autopilot_24_7: boolean;
  model: string;
  competitors?: string[];
  brand_keywords?: string[];
};

export type DailyBrief = {
  headline: string;
  action: string;
  why: string;
  insights: string[];
  day_key: string;
  generated_at?: string;
};

export type WeeklyReport = {
  week_key: string;
  title: string;
  summary?: string;
  highlights: string[];
  top_deliverables: string[];
  next_week_focus: string;
  wins?: string[];
  risks?: string[];
  next_week_priorities?: string[];
};

export type BrainFeedItem = {
  id: number;
  agent_id: string;
  asset_type: string;
  title: string;
  body: Record<string, unknown>;
  created_at: string;
};

export type AgentRunResult = {
  ok: boolean;
  error?: string;
  needs_integrations?: string[];
  agent_id?: string;
  agent_name?: string;
  summary?: string;
  deliverable?: Record<string, unknown>;
  executions?: { channel: string; ok: boolean; message?: string }[];
  intel_used?: Record<string, number | boolean>;
};

export type RunAllResult = {
  ok: boolean;
  ran: number;
  skipped: number;
  results: { agent_id: string; name: string; summary?: string }[];
  errors: { agent_id: string; error?: string }[];
};

export async function getBrainStatus(): Promise<BrainStatus> {
  return apiFetch<BrainStatus>("/api/v1/brain/status");
}

export async function getBrainAgents(): Promise<{ agents: BrainAgent[] }> {
  return apiFetch<{ agents: BrainAgent[] }>("/api/v1/brain/agents");
}

export async function getDailyBrief(refresh = false): Promise<DailyBrief> {
  return apiFetch<DailyBrief>(`/api/v1/brain/daily-brief?refresh=${refresh}`);
}

export async function getWeeklyReport(refresh = false): Promise<WeeklyReport> {
  return apiFetch<WeeklyReport>(`/api/v1/brain/weekly-report?refresh=${refresh}`);
}

export async function getBrainFeed(limit = 30): Promise<{ items: BrainFeedItem[] }> {
  return apiFetch<{ items: BrainFeedItem[] }>(`/api/v1/brain/feed?limit=${limit}`);
}

export async function triggerBrainLearn(force = false): Promise<{ ok: boolean }> {
  return apiFetch<{ ok: boolean }>(`/api/v1/brain/learn?force=${force}`, { method: "POST" });
}

export async function runBrainAgent(agentId: string): Promise<AgentRunResult> {
  return apiFetch<AgentRunResult>(`/api/v1/brain/agents/${agentId}/run`, { method: "POST" });
}

export async function runAllBrainAgents(): Promise<RunAllResult> {
  return apiFetch<RunAllResult>("/api/v1/brain/run-all", { method: "POST" });
}

export async function runMorningCycle(): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>("/api/v1/brain/morning-cycle", { method: "POST" });
}

export type BrainConfig = {
  competitors: string[];
  brand_keywords: string[];
  auto_run_daily: boolean;
  last_auto_run_key: string;
  enabled_agents: string[];
};

export async function getBrainConfig(): Promise<BrainConfig> {
  return apiFetch<BrainConfig>("/api/v1/brain/config");
}

export async function updateBrainConfig(patch: Partial<BrainConfig>): Promise<BrainConfig> {
  return apiFetch<BrainConfig>("/api/v1/brain/config", {
    method: "PATCH",
    body: JSON.stringify(patch),
  });
}

export async function ingestBrainUrl(url: string): Promise<{ ok: boolean; summary: string }> {
  return apiFetch<{ ok: boolean; summary: string }>("/api/v1/brain/ingest-url", {
    method: "POST",
    body: JSON.stringify({ url }),
  });
}
